File: tools/friday-launcher/confirm_darwin.go
//go:build darwin

package main

// confirmQuit shows the macOS Quit-confirmation dialog and returns
// true if the user clicked "Quit", false if "Cancel" (or dismissed
// via Esc).
//
// We use osascript here (via showStartupErrorDialog) — NOT cgo
// NSAlert — because the systray click handler runs on a Go
// goroutine, not the main thread, and macOS NSWindow/NSAlert APIs
// throw NSInternalInconsistencyException ("NSWindow should only be
// instantiated on the main thread!") when invoked off the main
// thread. Crash observed 2026-04-28: the launcher segfaulted on
// Quit, leaving every supervised process orphaned. osascript
// spawns AppleScript which has its own NSApp, no interaction with
// ours.
//
// Decision #2: explicit confirmation in front of the destructive
// "shut down all services" path. Cmd+Q (Decision #13) bypasses this
// — power-user signal, honor it immediately.
func confirmQuit() bool {
	const buttonQuit = "Quit"
	const buttonCancel = "Cancel"
	clicked := showStartupErrorDialog(
		"Quit Friday Studio?",
		"Friday Studio will stop all running services and shut down. "+
			"This may take up to 30 seconds.",
		[]string{buttonCancel, buttonQuit},
	)
	return clicked == buttonQuit
}

File: tools/friday-launcher/confirm_darwin.m
// confirm_darwin.m — cgo Objective-C shim for the Quit confirmation
// dialog. Called from confirmQuit() in confirm_darwin.go (which is
// invoked from the systray Quit menu-item click handler).
//
// Why NSAlert here (and not osascript): by the time the user clicks
// the tray's Quit menu item, NSApp is up — systray.Run has been
// running. NSAlert.runModal needs an NSApp; native dialog gives us
// nicer styling than osascript would. Pre-flight + bind-failure run
// BEFORE systray.Run and use osascript instead (see
// preflight_dialog_darwin.go).

#import <Cocoa/Cocoa.h>
#include <stdbool.h>

// friday_confirm_quit shows a modal NSAlert with two buttons (Quit /
// Cancel) and returns true iff the user clicked Quit. Synchronous —
// the systray click handler runs on the main thread, so blocking
// here is fine.
//
// Default button is the first one added ("Quit") — clicking through
// with Enter shuts down. macOS convention is "destructive default
// is OK if it's framed as a confirmation"; the tray's Quit menu
// item is itself an explicit user action, so re-confirming with
// Enter is the path of least friction.
bool friday_confirm_quit(void) {
    @autoreleasepool {
        NSAlert *alert = [[NSAlert alloc] init];
        [alert setMessageText:@"Quit Friday Studio?"];
        [alert setInformativeText:
            @"Friday Studio will stop all running services and "
             "shut down. This may take up to 30 seconds."];
        [alert setAlertStyle:NSAlertStyleWarning];
        [alert addButtonWithTitle:@"Quit"];
        [alert addButtonWithTitle:@"Cancel"];

        NSModalResponse response = [alert runModal];
        return response == NSAlertFirstButtonReturn;
    }
}
